// car_rental/routes.js
var mongoose = require('mongoose');

require('./models/Car.js');
require('./models/Rental.js');
require('./models/Model.js');
require('./models/Make.js');
require('./models/Customer.js');

var exports = module.exports = { 
    build: function (app, resolveUrl) {
        app.get("/", function(req, res) {
            res.json({
                applicationName : "Car rental",
                entities : resolveUrl("entities")
            });
        });
        
        app.get("/entities", function(req, res) {
            res.json([
                {
                   fullName : "car_rental.Car",
                   extentUri : resolveUrl("entities/car_rental.Car/instances"),
                   uri : resolveUrl("entities/car_rental.Car")
                }
                ,
                 {
                   fullName : "car_rental.Rental",
                   extentUri : resolveUrl("entities/car_rental.Rental/instances"),
                   uri : resolveUrl("entities/car_rental.Rental")
                }
                ,
                 {
                   fullName : "car_rental.Model",
                   extentUri : resolveUrl("entities/car_rental.Model/instances"),
                   uri : resolveUrl("entities/car_rental.Model")
                }
                ,
                 {
                   fullName : "car_rental.Make",
                   extentUri : resolveUrl("entities/car_rental.Make/instances"),
                   uri : resolveUrl("entities/car_rental.Make")
                }
                ,
                 {
                   fullName : "car_rental.Customer",
                   extentUri : resolveUrl("entities/car_rental.Customer/instances"),
                   uri : resolveUrl("entities/car_rental.Customer")
                }
            ]);
        });
        // routes for car_rental.Car
        app.get("/entities/car_rental.Car", function(req, res) {
            res.json({
                extentUri: resolveUrl('entities/car_rental.Car/instances')
            });
        });
        app.get("/entities/car_rental.Car/instances", function(req, res) {
            return mongoose.model('car_rental.Car').find().then(function(block) {
                res.json({
                    offset: 0,
                    length: block.size(),
                    contents: block
                });
            });
        });
        app.get("/entities/car_rental.Car/template", function(req, res) {
            res.json(new Car());
        });
        app.post("/entities/car_rental.Car/instances", function(req, res) {
            var instanceData = req.body;
            var newCar = new Car();
            for (var p in instanceData) {
                newCar[p] = instanceData[p]; 
            }
            newCar.save(function(err, created) {
                if (err) {
                    respondWithError(req, err);
                } else {
                    res.json(created);    
                }
            });
        });
        
        
        // routes for car_rental.Rental
        app.get("/entities/car_rental.Rental", function(req, res) {
            res.json({
                extentUri: resolveUrl('entities/car_rental.Rental/instances')
            });
        });
        app.get("/entities/car_rental.Rental/instances", function(req, res) {
            return mongoose.model('car_rental.Rental').find().then(function(block) {
                res.json({
                    offset: 0,
                    length: block.size(),
                    contents: block
                });
            });
        });
        app.get("/entities/car_rental.Rental/template", function(req, res) {
            res.json(new Rental());
        });
        app.post("/entities/car_rental.Rental/instances", function(req, res) {
            var instanceData = req.body;
            var newRental = new Rental();
            for (var p in instanceData) {
                newRental[p] = instanceData[p]; 
            }
            newRental.save(function(err, created) {
                if (err) {
                    respondWithError(req, err);
                } else {
                    res.json(created);    
                }
            });
        });
        
        
        // routes for car_rental.Model
        app.get("/entities/car_rental.Model", function(req, res) {
            res.json({
                extentUri: resolveUrl('entities/car_rental.Model/instances')
            });
        });
        app.get("/entities/car_rental.Model/instances", function(req, res) {
            return mongoose.model('car_rental.Model').find().then(function(block) {
                res.json({
                    offset: 0,
                    length: block.size(),
                    contents: block
                });
            });
        });
        app.get("/entities/car_rental.Model/template", function(req, res) {
            res.json(new Model());
        });
        app.post("/entities/car_rental.Model/instances", function(req, res) {
            var instanceData = req.body;
            var newModel = new Model();
            for (var p in instanceData) {
                newModel[p] = instanceData[p]; 
            }
            newModel.save(function(err, created) {
                if (err) {
                    respondWithError(req, err);
                } else {
                    res.json(created);    
                }
            });
        });
        
        
        // routes for car_rental.Make
        app.get("/entities/car_rental.Make", function(req, res) {
            res.json({
                extentUri: resolveUrl('entities/car_rental.Make/instances')
            });
        });
        app.get("/entities/car_rental.Make/instances", function(req, res) {
            return mongoose.model('car_rental.Make').find().then(function(block) {
                res.json({
                    offset: 0,
                    length: block.size(),
                    contents: block
                });
            });
        });
        app.get("/entities/car_rental.Make/template", function(req, res) {
            res.json(new Make());
        });
        app.post("/entities/car_rental.Make/instances", function(req, res) {
            var instanceData = req.body;
            var newMake = new Make();
            for (var p in instanceData) {
                newMake[p] = instanceData[p]; 
            }
            newMake.save(function(err, created) {
                if (err) {
                    respondWithError(req, err);
                } else {
                    res.json(created);    
                }
            });
        });
        
        
        // routes for car_rental.Customer
        app.get("/entities/car_rental.Customer", function(req, res) {
            res.json({
                extentUri: resolveUrl('entities/car_rental.Customer/instances')
            });
        });
        app.get("/entities/car_rental.Customer/instances", function(req, res) {
            return mongoose.model('car_rental.Customer').find().then(function(block) {
                res.json({
                    offset: 0,
                    length: block.size(),
                    contents: block
                });
            });
        });
        app.get("/entities/car_rental.Customer/template", function(req, res) {
            res.json(new Customer());
        });
        app.post("/entities/car_rental.Customer/instances", function(req, res) {
            var instanceData = req.body;
            var newCustomer = new Customer();
            for (var p in instanceData) {
                newCustomer[p] = instanceData[p]; 
            }
            newCustomer.save(function(err, created) {
                if (err) {
                    respondWithError(req, err);
                } else {
                    res.json(created);    
                }
            });
        });
    }
};
